### PR TITLE
feat(web): multi tenancy

### PR DESCRIPTION
**Pull request by RhysSullivan**
# Description

In order for this to be complete, the following checks must be complete:

- [ ] Canonical URLs link to correct url
- [ ] Replace sitemap urls from answeroverflow.com to domain.com
- [ ] Store server owner domains in edge cache
- [ ] Add field for server owner domains in server
- [ ] Favicon for server owner pages needs to be their discord icon
- [ ] Send server owner domains to client for canoncial urls
- [ ] Redirect to answeroverflow.com if expired
- [ ] Visiting a URL directly should redirect to tenant site
- [ ] Query params transfer correctly
- [ ] Auth needs to redirect back to correct page
- [ ] Auth cookie needs to be owned only by answeroverflow.com
- [ ] Replace link components w/ tenant links
- [ ] Sign out on tenant site
- [ ] Sign in on tenant site
- [ ] Searches from a tenant domain need to have their query string rewritten / on the API level be filtered only to that community

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works 
- [ ] I have updated .env.example if I added a new environment variable
- [ ] My PR title follows the [semantic commits](https://www.conventionalcommits.org/en/v1.0.0/) style

### CODE REVIEW

File: apps/main-site/src/pages/api/auth/[...nextauth].ts [hunk 1]
```diff
@@ -1,5 +1,32 @@
 import NextAuth from 'next-auth';
-
 import { authOptions } from '@answeroverflow/auth';
+import type { NextApiRequest, NextApiResponse } from 'next';
+
+// @see https://nextjs.org/docs/api-routes/introduction
+export default async function handler(
```
Comment (RhysSullivan) [inline apps/main-site/src/pages/api/auth/[...nextauth].ts:6]: Swap to doing in middleware

File: apps/main-site/src/pages/api/trpc/[trpc].ts [hunk 2]
```diff
@@ -15,11 +15,12 @@ export default async function handler(
 	req: NextApiRequest,
 	res: NextApiResponse,
 ) {
+	// TODO: Only run this if not on the main domain
 	// Only enable CORS in development for accessing through Storybook
 	if (process.env.NODE_ENV !== 'production') {
 		// Modify `req` and `res` objects here
 		// In this case, we are enabling CORS
-		res.setHeader('Access-Control-Allow-Origin', 'http://localhost:6006');
+		res.setHeader('Access-Control-Allow-Origin', 'http://tenant:3001');
```
Comment (RhysSullivan) [inline apps/main-site/src/pages/api/trpc/[trpc].ts:23]: Move to middelware

File: packages/auth/src/auth-options.ts [hunk 3]
```diff
@@ -13,6 +13,9 @@ import {
 	updateProviderAuthToken,
 } from '@answeroverflow/db';
 
+// TODO: What should this be?
+const VERCEL_DEPLOYMENT = !!process.env.VERCEL_URL;
```
Comment (RhysSullivan) [inline packages/auth/src/auth-options.ts:17]: revisit

File: packages/auth/src/auth-options.ts [hunk 4]
```diff
@@ -25,6 +28,19 @@ export const authOptions: NextAuthOptions = {
 		}),
 		// ...add more providers here
 	],
+	cookies: {
+		sessionToken: {
+			name: `${VERCEL_DEPLOYMENT ? '__Secure-' : ''}next-auth.session-token`,
+			options: {
+				httpOnly: true,
+				sameSite: 'none',
```
Comment (RhysSullivan) [inline packages/auth/src/auth-options.ts:36]: ensure is secure

File: packages/auth/src/auth-options.ts [hunk 5]
```diff
@@ -25,6 +28,19 @@ export const authOptions: NextAuthOptions = {
 		}),
 		// ...add more providers here
 	],
+	cookies: {
+		sessionToken: {
+			name: `${VERCEL_DEPLOYMENT ? '__Secure-' : ''}next-auth.session-token`,
```
Comment (RhysSullivan) [inline packages/auth/src/auth-options.ts:33]: revisit

File: packages/ui/src/components/pages/CommunityPage.tsx [hunk 6]
```diff
@@ -85,7 +85,23 @@ export const CommunityPage = ({ server, channels }: CommunityPageData) => {
 	const [selectedChannelId, setSelectedChannelId] = useState<null | string>(
 		channels.at(0)?.channel.id ?? null,
 	);
-
+	useEffect(() => {
```
Comment (RhysSullivan) [inline packages/ui/src/components/pages/CommunityPage.tsx:88]: remove

File: packages/ui/src/utils/trpc.ts [hunk 7]
```diff
@@ -6,17 +6,19 @@ import type { AppRouter } from '@answeroverflow/api';
 import { transformer } from '@answeroverflow/api/transformer';
 import { createTRPCReact } from '@trpc/react-query';
 
+// TODO: Fix
 const getBaseUrl = () => {
```
Comment (RhysSullivan) [inline packages/ui/src/utils/trpc.ts:10]: revisit

File: packages/ui/src/utils/trpc.ts [hunk 8]
```diff
@@ -6,17 +6,19 @@ import type { AppRouter } from '@answeroverflow/api';
...
-	if (typeof window !== 'undefined') return ''; // browser should use relative url
+	if (typeof window !== 'undefined') return `http://localhost:3000`; // browser should use relative url
 	if (process.env.VERCEL_URL) return `https://${process.env.VERCEL_URL}`; // SSR should use vercel url
-	return `http://localhost:${process.env.PORT ?? 3000}`; // dev SSR should use localhost
+	return `http://localhost:3000`; // dev SSR should use localhost
```
Comment (RhysSullivan) [inline packages/ui/src/utils/trpc.ts:13]: revisit

File: packages/ui/src/utils/trpc.ts [hunk 9]
```diff
@@ -25,6 +27,14 @@ const nextTRPC = () =>
 					}),
 					httpBatchLink({
 						url: `${getBaseUrl()}/api/trpc`,
+						fetch(url, opts) {
+							return fetch(url, {
+								...opts,
+								// We're doing this for multi tenancy
+								// TODO: Maybe don't include if on the main domain?
+								credentials: 'include',
```
Comment (RhysSullivan) [inline packages/ui/src/utils/trpc.ts:35]: revisit

**Comment by RhysSullivan** [comment]
Good proof of concept, needs polish

**Comment by mrhappyma** [comment]
sorry i'm using you as an example :)

**Pull request closed without merging**